### ISSUE TITLE
kairos reset fails on Jetson Orin device

### ISSUE THREAD

**Issue by nianyush** [kairos-io/kairos#1723]
**Kairos version:**

NAME="Ubuntu"
VERSION="20.04.6 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.6 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal
KAIROS_NAME="kairos-core-ubuntu-20-lts-arm-nvidia-jetson-agx-orin"
KAIROS_VERSION="v2.3.2"
KAIROS_ID="kairos"
KAIROS_ID_LIKE="kairos-core-ubuntu-20-lts-arm-nvidia-jetson-agx-orin"
KAIROS_VERSION_ID="v2.3.2"
KAIROS_PRETTY_NAME="kairos-core-ubuntu-20-lts-arm-nvidia-jetson-agx-orin v2.3.2"
KAIROS_BUG_REPORT_URL="https://github.com/kairos-io/kairos/issues"
KAIROS_HOME_URL="https://github.com/kairos-io/kairos"
KAIROS_IMAGE_REPO="quay.io/kairos/core-ubuntu-20-lts-arm-nvidia-jetson-agx-orin"
KAIROS_IMAGE_LABEL="latest"
KAIROS_GITHUB_REPO="kairos-io/kairos"
KAIROS_VARIANT="core"
KAIROS_FLAVOR="ubuntu-20-lts-arm-nvidia-jetson-agx-orin"
**CPU architecture, OS, and Version:**

Linux cos-recovery 5.10.104-tegra #1 SMP PREEMPT Sun Mar 19 07:55:28 PDT 2023 aarch64 aarch64 aarch64 GNU/Linux
**Describe the bug**

reset failed and went into cos-recovery
**To Reproduce**

Flash the device with this image quay.io/kairos/core-ubuntu-20-lts-arm-nvidia-jetson-agx-orin:v2.3.2
choose reset in the grub menu during startup
**Expected behavior**

**Logs**

[FAILED] Failed to start kairos reset.
See 'systemctl status kairos-reset.service' for details.
[FAILED] Failed to start iSCSI initiator daemon (iscsid).
See 'systemctl status iscsid.service' for details.
[  OK  ] Reached target Remote File Systems (Pre).
[  OK  ] Reached target Remote File Systems.
         Starting Availability of block devices...
         Starting Permit User Sessions...
[  OK  ] Finished Availability of block devices.
[  OK  ] Finished Permit User Sessions.
[  OK  ] Started Serial Getty on ttyTCU0.
[  OK  ] Reached target Login Prompts.
[  OK  ] Started UART on ttyTHS0.
         Starting Configure bootloader service...
         Starting Set console scheme...

Ubuntu 20.04.6 LTS cos-recovery ttyTCU0

You are booting from recovery mode. Run 'kairos-agent reset' to reset the system to
 or 'kairos-agent upgrade' to upgrade the active partition

cos-recovery login:

root@cos-recovery:~# systemctl status kairos-reset.service
● kairos-reset.service - kairos reset
     Loaded: loaded (/etc/systemd/system/kairos-reset.service; enabled; vendor preset: enabled)
     Active: failed (Result: exit-code) since Tue 2023-08-15 20:21:58 UTC; 13min ago
    Process: 1274 ExecStart=/usr/bin/kairos-agent reset (code=exited, status=1/FAILURE)
   Main PID: 1274 (code=exited, status=1/FAILURE)

Mar 27 17:54:15 jetson-6 systemd[1]: Starting kairos reset...
Aug 15 20:21:51 cos-recovery /usr/local/bin/agent-provider-stylus[2862]: time="2023-08-15T20:21:51Z" level=info msg="Starting agent-provider-stylus version v4.0.0-alpha13"
Aug 15 20:21:58 cos-recovery systemd[1]: kairos-reset.service: Main process exited, code=exited, status=1/FAILURE
Aug 15 20:21:58 cos-recovery systemd[1]: kairos-reset.service: Failed with result 'exit-code'.
Aug 15 20:21:58 cos-recovery systemd[1]: Failed to start kairos reset.
**Additional context**

**Comment by mudler** [comment]
@nianyush can you try manually and provide the full logs?  from the snippet seems to have no output

To try manually, just boot into recovery and try to run kairos-agent reset directly from there

**Comment by mudler** [comment]
A manual `kairos-agent reset` works here from recovery:

```
┌────────────────────────────────────────────────────────────────────┐  
| Welcome to kairos!                                                 |      
| The node will automatically reset its state in a few.              |  
|                                                                    |  
| Press any key to abort this process. To restart run 'kairos reset'.|  
|                                                                    |     
| Starting in 60 seconds...                                          |  
└──────────────────────────────────────────────────────────── Reset ─┘     
warning: skipping /oem/grubenv (extension).                                                                                                                                                                          
WARNING: jsonschema: '' does not validate with file:///root/schema.json#/required: missing properties: 'users'
INFO[2023-08-16T12:51:15Z] kairos-agent version v2.1.12                 
WARN[2023-08-16T12:51:15Z] failed reading installation state: open /run/initramfs/cos-state/state.yaml: no such file or directory 
WARN[2023-08-16T12:51:15Z] no Persistent partition found                 
INFO[2023-08-16T12:51:15Z] Unmounting disk partitions                                          
INFO[2023-08-16T12:51:15Z] Formatting 'state' partition                  
INFO[2023-08-16T12:51:24Z] Mounting disk partitions                                                                                                                                                                  
INFO[2023-08-16T12:51:25Z] Running before-reset hook                     
INFO[2023-08-16T12:51:25Z] Running stage: before-reset.before                                    
INFO[2023-08-16T12:51:25Z] Done executing stage 'before-reset.before'                                                                                                                                                
INFO[2023-08-16T12:51:25Z] Running stage: before-reset                                                                                                                                                               
INFO[2023-08-16T12:51:25Z] Done executing stage 'before-reset'                                  
INFO[2023-08-16T12:51:25Z] Running stage: before-reset.after             
INFO[2023-08-16T12:51:25Z] Done executing stage 'before-reset.after'    
INFO[2023-08-16T12:51:25Z] Running stage: before-reset.before           
INFO[2023-08-16T12:51:25Z] Done executing stage 'before-reset.before'   
INFO[2023-08-16T12:51:25Z] Running stage: before-reset                  
INFO[2023-08-16T12:51:25Z] Done executing stage 'before-reset'          
INFO[2023-08-16T12:51:25Z] Running stage: before-reset.after             
INFO[2023-08-16T12:51:25Z] Done executing stage 'before-reset.after'    
INFO[2023-08-16T12:51:25Z] Some errors found but were ignored. Enable --strict mode to fail on those or --debug to see them in the log 
WARN[2023-08-16T12:51:25Z] 3 errors occurred:                                                             
        * yaml: unmarshal errors:                                                                                                                                                                                    
  line 1: cannot unmarshal !!str `/usr/lo...` into schema.YipConfig
        * yaml: unmarshal errors:
  line 1: cannot unmarshal !!str `/usr/lo...` into schema.YipConfig
        * yaml: unmarshal errors:
  line 1: cannot unmarshal !!str `/usr/lo...` into schema.YipConfig
                                 
INFO[2023-08-16T12:51:25Z] Copying /run/initramfs/cos-state/cOS/recovery.img source... 
INFO[2023-08-16T12:52:41Z] Finished copying /run/initramfs/cos-state/cOS/recovery.img into /run/cos/state/cOS/active.img 
INFO[2023-08-16T12:52:44Z] Using grub config dir /run/cos/active/etc/cos/grub.cfg 
INFO[2023-08-16T12:52:44Z] Copying grub contents from /run/cos/active/etc/cos/grub.cfg to /run/cos/state/grub2/grub.cfg 
INFO[2023-08-16T12:52:44Z] Generating grub files for efi on /dev/mmcblk0 
INFO[2023-08-16T12:52:45Z] Running after-reset-chroot hook              
INFO[2023-08-16T12:52:45Z] Running stage: after-reset-chroot.before     
INFO[2023-08-16T12:52:45Z] Done executing stage 'after-reset-chroot.before' 
INFO[2023-08-16T12:52:45Z] Running stage: after-reset-chroot            
```

@nianyush what's your setup? do you have COS_PERSISTENT in the eMMC?

**Comment by mudler** [comment]
there is something odd tho - I have a COS_PERSISTENT but can't find it apparently:

```
WARN[2023-08-16T12:51:15Z] no Persistent partition found                 
```
however, doesn't block from the reset operation that seems to complete successfully here:

```
INFO[2023-08-16T12:52:45Z] Running stage: after-reset-chroot.after      
INFO[2023-08-16T12:52:45Z] Done executing stage 'after-reset-chroot.after' 
INFO[2023-08-16T12:52:45Z] Running stage: after-reset-chroot.before     
INFO[2023-08-16T12:52:45Z] Done executing stage 'after-reset-chroot.before' 
INFO[2023-08-16T12:52:45Z] Running stage: after-reset-chroot            
INFO[2023-08-16T12:52:45Z] Done executing stage 'after-reset-chroot'    
INFO[2023-08-16T12:52:45Z] Running stage: after-reset-chroot.after      
INFO[2023-08-16T12:52:45Z] Done executing stage 'after-reset-chroot.after' 
INFO[2023-08-16T12:52:45Z] Setting default grub entry to Kairos         
INFO[2023-08-16T12:52:50Z] Copying /run/cos/state/cOS/active.img source... 
INFO[2023-08-16T12:53:11Z] Finished copying /run/cos/state/cOS/active.img into /run/cos/state/cOS/passive.img 
INFO[2023-08-16T12:53:37Z] Running after-reset hook                      
INFO[2023-08-16T12:53:37Z] Running stage: after-reset.before            
INFO[2023-08-16T12:53:37Z] Done executing stage 'after-reset.before'    
INFO[2023-08-16T12:53:37Z] Running stage: after-reset                    
INFO[2023-08-16T12:53:37Z] Processing stage step 'Mount state'. ( commands: 1, files: 0, ... ) 
INFO[2023-08-16T12:53:37Z] Command output:                               
INFO[2023-08-16T12:53:37Z] Processing stage step 'Hook boot assessment grub configuration'. ( commands: 1, files: 0, ... ) 
INFO[2023-08-16T12:53:37Z] Command output:                               
INFO[2023-08-16T12:53:37Z] Processing stage step 'Grub branding'. ( commands: 1, files: 0, ... ) 
INFO[2023-08-16T12:53:37Z] Command output: '/etc/kairos/branding/grubmenu.cfg' -> '/tmp/mnt/STATE/grubmenu' 
INFO[2023-08-16T12:53:37Z] Processing stage step 'Add boot assessment grub configuration'. ( commands: 0, files: 1, ... ) 
INFO[2023-08-16T12:53:37Z] Processing stage step 'umount state'. ( commands: 1, files: 0, ... ) 
INFO[2023-08-16T12:53:37Z] Command output:                               
INFO[2023-08-16T12:53:37Z] Done executing stage 'after-reset'           
INFO[2023-08-16T12:53:37Z] Running stage: after-reset.after             
INFO[2023-08-16T12:53:37Z] Done executing stage 'after-reset.after'     
INFO[2023-08-16T12:53:37Z] Running stage: after-reset.before            
INFO[2023-08-16T12:53:37Z] Done executing stage 'after-reset.before'    
INFO[2023-08-16T12:53:37Z] Running stage: after-reset                    
INFO[2023-08-16T12:53:37Z] Done executing stage 'after-reset'           
INFO[2023-08-16T12:53:37Z] Running stage: after-reset.after             
INFO[2023-08-16T12:53:37Z] Done executing stage 'after-reset.after'     
INFO[2023-08-16T12:53:37Z] Some errors found but were ignored. Enable --strict mode to fail on those or --debug to see them in the log 
WARN[2023-08-16T12:53:37Z] 3 errors occurred:
        * yaml: unmarshal errors:
  line 1: cannot unmarshal !!str `/usr/lo...` into schema.YipConfig
        * yaml: unmarshal errors:
  line 1: cannot unmarshal !!str `/usr/lo...` into schema.YipConfig
        * yaml: unmarshal errors:
  line 1: cannot unmarshal !!str `/usr/lo...` into schema.YipConfig
  
INFO[2023-08-16T12:53:38Z] Unmounting disk partitions                    
 INFO  Rebooting in 60 seconds, press Enter to abort...
 INFO  Rebooting node

Session terminated, killing shell... ...killed.
Terminated
kairos@cos-recovery:~$ 
```

**Comment by venkatnsrinivasan** [comment]
i think he has a COS_PERSISTENT parition. Manual reset gave the partition in use error and failed. @nianyush you can try that and see . But the runtime kairos-reset exits with status 1  . We cant seem to find any error logs.

**Comment by nianyush** [comment]
@venkatnsrinivasan @mudler I was using usb persistent partition

**Comment by venkatnsrinivasan** [comment]
Yes @mudler we dont add it it xml if we are using external usb.

**Comment by nianyush** [comment]
@mudler
Output of `blkid`
```
root@cos-recovery:~# blkid
/dev/mmcblk0p16: LABEL="COS_RECOVERY" UUID="983b55a6-b6a3-41ef-bb8f-bf0a2bf8cc59" TYPE="ext4" PARTLABEL="COS_RECOVERY" PARTUUID="32f2f99e-e5e7-4e3b-995a-6b290b2cb224"
/dev/mmcblk0p18: LABEL="COS_OEM" UUID="5ea36634-6a3b-4878-bf69-952c3550a67e" TYPE="ext2" PARTLABEL="COS_OEM" PARTUUID="7a3629c1-f15e-4db5-a38c-7e301a413a66"
/dev/loop0: LABEL="COS_SYSTEM" UUID="c33022f5-d09a-4543-8448-71126affca9c" TYPE="ext2"
/dev/mmcblk0p2: PARTLABEL="A_kernel" PARTUUID="0ae93b85-ddc2-4ddf-96bc-c0636fe9b26f"
/dev/mmcblk0p3: PARTLABEL="A_kernel-dtb" PARTUUID="47a83a02-5179-41be-9370-394de3b3072c"
/dev/mmcblk0p4: PARTLABEL="A_reserved_on_user" PARTUUID="24dda738-548d-4533-92ed-dc374fe8f30b"
/dev/mmcblk0p5: PARTLABEL="B_kernel" PARTUUID="755e5c65-163a-4f31-9259-272884997b15"
/dev/mmcblk0p6: PARTLABEL="B_kernel-dtb" PARTUUID="39201bf8-2e02-430b-b735-5e03ef61e916"
/dev/mmcblk0p7: PARTLABEL="B_reserved_on_user" PARTUUID="468cccc8-9059-48a7-9b7b-354a046d7a76"
/dev/mmcblk0p8: PARTLABEL="recovery" PARTUUID="17abce33-f57f-4077-a57e-2c092ab3175c"
/dev/mmcblk0p9: PARTLABEL="recovery-dtb" PARTUUID="71bf326c-3ae7-4f9e-8909-8536f26da87c"
/dev/mmcblk0p10: SEC_TYPE="msdos" LABEL_FATBOOT="COS_GRUB" LABEL="COS_GRUB" UUID="05E5-8F60" TYPE="vfat" PARTLABEL="esp" PARTUUID="4d7e18a9-c5df-4a45-a157-5b6cab522615"
/dev/mmcblk0p11: PARTLABEL="recovery_alt" PARTUUID="6c041758-c834-4994-8f06-2e4191bee110"
/dev/mmcblk0p12: PARTLABEL="recovery-dtb_alt" PARTUUID="4ec81cc1-f421-490a-a0a6-d51c27792644"
/dev/mmcblk0p13: PARTLABEL="esp_alt" PARTUUID="583c0a5b-fff2-44fc-ab12-a25954265c11"
/dev/mmcblk0p14: PARTLABEL="reserved" PARTUUID="18082df4-4862-4d00-8388-4528bdfa945e"
/dev/mmcblk0p15: PARTLABEL="esp" PARTUUID="35a7d8bc-041f-427b-8167-0f55efa6534d"
/dev/mmcblk0p17: LABEL="COS_STATE" UUID="c8d2996f-7b75-4bc6-893a-b3cd68cf516f" TYPE="ext4" PARTLABEL="COS_STATE" PARTUUID="4dda20ce-63e2-4ff0-bd99-5a217739581b"
/dev/sda1: LABEL="COS_PERSISTENT" UUID="e8808105-d1f7-4dbb-b0dc-d26989d56a53" TYPE="ext4" PARTLABEL="Main Data Partition" PARTUUID="6a07ece1-4f6a-493d-8e86-6ab19d500ed0"
/dev/sda2: PARTLABEL="primary" PARTUUID="fcd92c32-1897-458f-a8fc-3c14affd94ac"
/dev/zram0: UUID="e51614dd-aa7a-4620-82c4-67e7add51bcf" TYPE="swap"
/dev/zram1: UUID="b9a605f0-d497-4c32-96fc-cff50e30e2e4" TYPE="swap"
/dev/zram2: UUID="42920f5b-25be-438d-8e99-dd3b330136d8" TYPE="swap"
/dev/zram3: UUID="e975f6c4-c281-43a6-8d0d-e7a24021e071" TYPE="swap"
/dev/zram4: UUID="e2e8037b-ffa5-4dd0-9e5f-e1a9204db5ea" TYPE="swap"
/dev/zram5: UUID="c4cce075-8a9e-47be-a8a1-323216cc2589" TYPE="swap"
/dev/zram6: UUID="4f920fc2-7439-4b82-8068-d5b1fae90595" TYPE="swap"
/dev/zram7: UUID="3b3a5a7e-d6ce-4f50-b5aa-c878dc6d6b3b" TYPE="swap"
/dev/zram8: UUID="aa9eaa18-9b49-46bc-9e5c-0f0537613cd3" TYPE="swap"
/dev/zram9: UUID="085051ef-ea2c-4365-9412-98d8609e0abb" TYPE="swap"
/dev/zram10: UUID="12550370-85ef-47b2-98a2-25ea0d460887" TYPE="swap"
/dev/zram11: UUID="38bbb2fe-46d2-45c0-9968-d47595abeb5b" TYPE="swap"
```
Output of `kairos-agent state`
```
root@cos-recovery:~# kairos-agent state
uuid: 1985b743ecf04b25a9c9cbfb9525ce4b-cos-recovery
persistent:
    mounted: true
    name: /dev/sda1
    label: Main\x20Data\x20Partition
    filesystemlabel: COS_PERSISTENT
    mount_point: /run/mount/persistent
    size_bytes: 128042631168
    type: ext4
    read_only: false
    found: true
    uuid: 6a07ece1-4f6a-493d-8e86-6ab19d500ed0
recovery:
    mounted: true
    name: /dev/mmcblk0p16
    label: COS_RECOVERY
    filesystemlabel: COS_RECOVERY
    mount_point: /run/initramfs/cos-state
    size_bytes: 10485760000
    type: ext4
    read_only: true
    found: true
    uuid: 32f2f99e-e5e7-4e3b-995a-6b290b2cb224
oem:
    mounted: true
    name: /dev/mmcblk0p18
    label: COS_OEM
    filesystemlabel: COS_OEM
    mount_point: /oem
    size_bytes: 67108864
    type: ext2
    read_only: false
    found: true
    uuid: 7a3629c1-f15e-4db5-a38c-7e301a413a66
state:
    mounted: false
    name: /dev/mmcblk0p17
    label: COS_STATE
    filesystemlabel: COS_STATE
    mount_point: ""
    size_bytes: 14680064000
    type: ext4
    read_only: true
    found: true
    uuid: 4dda20ce-63e2-4ff0-bd99-5a217739581b
boot: recovery_boot
system:
    meta:
        version: 0.9.5
        timestamp: 2023-08-17T23:19:10.432427391Z
    node:
        hostname: cos-recovery
        machineid: 1985b743ecf04b25a9c9cbfb9525ce4b
        hypervisor: ""
        timezone: /UTC
    os:
        name: Ubuntu 20.04.6 LTS
        vendor: ubuntu
        version: "20.04"
        release: 20.04.6
        architecture: ""
    kernel:
        release: 5.10.104-tegra
        version: '#1 SMP PREEMPT Sun Mar 19 07:55:28 PDT 2023'
        architecture: aarch64
    product:
        name: Jetson AGX Orin
        vendor: Unknown
        version: Not Specified
        serial: "1420923016741"
    board:
        name: Jetson AGX Orin
        vendor: Unknown
        version: Not Specified
        serial: "1420923016741"
        assettag: 699-13701-0005-500 M.0
    chassis:
        type: 1
        vendor: Unknown
        version: Unknown
        serial: Unknown
        assettag: 699-13701-0005-500 M.0
    bios:
        vendor: EDK II
        version: 3.1-32827747
        date: 03/19/2023
    cpu:
        vendor: ""
        model: ARMv8 Processor rev 1 (v8l)
        speed: 2000
        cache: 0
        cpus: 0
        cores: 0
        threads: 12
    memory:
        type: ""
        speed: 0
        size: 0
    storage:
        - name: mmcblk0
          driver: mmcblk
          vendor: ""
          model: ""
          serial: ""
          size: 63
        - name: sda
          driver: sd
          vendor: SanDisk
          model: Extreme Pro
          serial: 51A45678C003
          size: 128
    network:
        - name: eth0
          driver: nvethernet
          macaddress: 48:b0:2d:e2:bc:af
          port: ""
          speed: 10000
        - name: wlP1p1s0
          driver: rtl88x2ce
          macaddress: b4:8c:9d:1c:50:bb
          port: ""
          speed: 0
kairos:
    flavor: ubuntu-20-lts-arm-nvidia-jetson-agx-orin
    version: v2.3.2
```
Output of

**Comment by mudler** [comment]
https://github.com/kairos-io/packages/pull/387 approach seems to work here, so until that is merged and a released here is a workaround.

Edit the recovery image and replace the content of `/system/oem/mount.yaml` with:

```yaml
## TODO: this is a workaround
## The orin packages are writing to /usr/local, which is mounted to COS_PERSISTENT.
## We probably should run this in immucore, overlaying the /usr/local of the image to COS_PERSISTENT.
## For the time being, doing it in the cloud configs.
#
### Note: This have the consequences of everything stored inside /usr/local to go inside the "local" directory inside the partition.
###       Usually this is not the case as there is no need of subtrees, but due to how overlayfs work the workdir and uppermount needs to be in the same filesystem.
stages:
  initramfs.before:
  - if: '[ ! -f "/run/cos/recovery_mode" ]' 
    commands:
    - umount /usr/local
    - |
       mkdir -p /run/mount/persistent && \
       mount /dev/disk/by-label/COS_PERSISTENT /run/mount/persistent && \
       mkdir -p /run/mount/persistent/work /run/mount/persistent/local && \
       mount -t overlay \
                  -o lowerdir=/usr/local,upperdir=/run/mount/persistent/local,workdir=/run/mount/persistent/work \
                  /usr/local
```

alternatively, you can try my custom image that contains that file already, but you should flash the device with it: `quay.io/mudler/tests:orin-nvidia`

**Comment by rishi-anand** [comment]
@mudler In which version, this fix will be released?

**Comment by mudler** [comment]
> @mudler In which version, this fix will be released?

this is planned for 2.4.0 that is planned for end of this month

**Comment by mudler** [comment]
Note this has not been validated yet on master - currently there are breaking changes.

**Comment by jimmykarily** [comment]
The fix is there on v2.4.0 so I'll close this now. Please re-open if something doesn't work. Thanks for reporting!
